File: src/data/courses/dsa/trie/trie-dfs-backtracking/boggle-board/code/java/BoggleBoard.java
import java.util.*;

class BoggleBoard {
    class TrieNode {
        TrieNode[] children = new TrieNode[26];
        String word = null; // Store word at end
    }
    
    private int[][] dirs = {{-1,-1},{-1,0},{-1,1},{0,-1},{0,1},{1,-1},{1,0},{1,1}}; // 8 directions
    
    public List<String> findWords(char[][] board, String[] words) {
        TrieNode root = new TrieNode();
        for (String word : words) insert(root, word);
        
        List<String> result = new ArrayList<>();
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                dfs(board, i, j, root, result);
            }
        }
        return result;
    }
    
    private void insert(TrieNode root, String word) {
        TrieNode node = root;
        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            if (node.children[idx] == null) node.children[idx] = new TrieNode();
            node = node.children[idx];
        }
        node.word = word; // Store complete word
    }
    
    private void dfs(char[][] board, int i, int j, TrieNode node, List<String> result) {
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length) return;
        
        char c = board[i][j];
        if (c == '#' || node.children[c - 'a'] == null) return; // Visited or no path
        
        node = node.children[c - 'a'];
        if (node.word != null) {
            result.add(node.word); // Found word
            node.word = null; // Avoid duplicates
        }
        
        board[i][j] = '#'; // Mark visited
        for (int[] dir : dirs) { // Explore 8 directions
            dfs(board, i + dir[0], j + dir[1], node, result);
        }
        board[i][j] = c; // Backtrack
    }
}

File: src/data/courses/dsa/trie/trie-dfs-backtracking/boggle-board/code/cpp/BoggleBoard.cpp
// ============================================================
// Solution 1: DFS per word separately
// Time: O(W * M * N * 4^L) | Space: O(L)
// ============================================================
#include <vector>
#include <string>
using namespace std;

class Solution1 {
    int m, n;
    bool dfs(vector<vector<char>>& board, string& word, int i, int j, int k) {
        if (k == (int)word.size()) return true;
        if (i < 0 || i >= m || j < 0 || j >= n || board[i][j] != word[k]) return false;
        char tmp = board[i][j]; board[i][j] = '#';
        bool found = dfs(board,word,i+1,j,k+1) || dfs(board,word,i-1,j,k+1) ||
                     dfs(board,word,i,j+1,k+1) || dfs(board,word,i,j-1,k+1);
        board[i][j] = tmp;
        return found;
    }
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        m = board.size(); n = board[0].size();
        vector<string> result;
        for (auto& w : words) {  // DFS per word — redundant grid traversal!
            bool found = false;
            for (int i = 0; i < m && !found; i++)
                for (int j = 0; j < n && !found; j++)
                    if (dfs(board, w, i, j, 0)) found = true;
            if (found) result.push_back(w);
        }
        return result;
    }
};

// ============================================================
// Solution 2: Optimal (Trie-based)
// ============================================================
#include <vector>
#include <string>
using namespace std;

class BoggleBoard {
    struct TrieNode {
        TrieNode* children[26] = {};
        string word = ""; // Store word at end
    };
    
    int dirs[8][2] = {{-1,-1},{-1,0},{-1,1},{0,-1},{0,1},{1,-1},{1,0},{1,1}}; // 8 directions
    
    void insert(TrieNode* root, string word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx]) node->children[idx] = new TrieNode();
            node = node->children[idx];
        }
        node->word = word; // Store complete word
    }
    
    void dfs(vector<vector<char>>& board, int i, int j, TrieNode* node, vector<string>& result) {
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size()) return;
        
        char c = board[i][j];
        if (c == '#' || !node->children[c - 'a']) return; // Visited or no path
        
        node = node->children[c - 'a'];
        if (!node->word.empty()) {
            result.push_back(node->word); // Found word
            node->word = ""; // Avoid duplicates
        }
        
        board[i][j] = '#'; // Mark visited
        for (auto& dir : dirs) { // Explore 8 directions
            dfs(board, i + dir[0], j + dir[1], node, result);
        }
        board[i][j] = c; // Backtrack
    }
    
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        TrieNode* root = new TrieNode();
        for (string& word : words) insert(root, word);
        
        vector<string> result;
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                dfs(board, i, j, root, result);
            }
        }
        return result;
    }
};
